pkg/plugins/events/eventgrid/eventgrid.go

```diff
@@ -0,0 +1,209 @@
+	newErr := errors.ErrorsWithScope(
+		"EventGrid.ListTopics",
+		map[string]interface{}{
+			"list": "topics",
+		},
+	)
+	//Set the topic page length
+	pageLength := int32(10)
+
+	ctx := context.Background()
+	results, err := s.topicClient.ListBySubscription(ctx, "", &pageLength)
+
+	if err != nil {
+		return nil, newErr(
+			codes.Internal,
+			"error listing by subscription",
+			err,
+		)
+	}
+
+	var topics []string
+
+	//Iterate over the topic pages adding their names to the topics slice
+	for results.NotDone() {
+		topicsList := results.Values()
+		for _, topic := range topicsList {
+			topics = append(topics, *topic.Name)
+		}
+		results.NextWithContext(ctx)
+	}
+
+	return topics, nil
+}
+
+func (s *EventGridEventService) GetTopicEndpoint(topicName string) (string, error) {
+	ctx := context.Background()
+	pageLength := int32(10)
+	results, err := s.topicClient.ListBySubscription(ctx, "", &pageLength)
+	if err != nil {
+		return "", fmt.Errorf(err.Error())
+	}
+
+	for results.NotDone() {
+		topicsList := results.Values()
+		for _, topic := range topicsList {
+			if *topic.Name == topicName {
+				return strings.TrimSuffix(strings.TrimPrefix(*topic.Endpoint, "https://"), "/api/events"), nil
+			}
+		}
+		results.Next()
+	}
+	return "", fmt.Errorf("topic with provided name could not be found")
+}
+
+func (s *EventGridEventService) NitricEventToEvent(topic string, event *events.NitricEvent) ([]eventgrid.Event, error) {
+	payload, err := json.Marshal(event.Payload)
+	if err != nil {
+		return nil, err
+	}
+	dataVersion := "1.0"
+	azureEvent := []eventgrid.Event{
+		{
+			ID:          &event.ID,
+			Data:        &payload,
+			EventType:   &event.PayloadType,
+			Subject:     &topic,
+			EventTime:   &date.Time{time.Now()},
+			DataVersion: &dataVersion,
+		},
+	}
+
+	return azureEvent, nil
+}
+
+func (s *EventGridEventService) Publish(topic string, event *events.NitricEvent) error {
+	newErr := errors.ErrorsWithScope(
+		"EventGrid.Publish",
+		map[string]interface{}{
+			"topic": topic,
+		},
+	)
+	ctx := context.Background()
+
+	if len(topic) == 0 {
+		return newErr(
+			codes.InvalidArgument,
+			"provided invalid topic",
+			fmt.Errorf(""),
+		)
+	}
+	if event == nil {
+		return newErr(
+			codes.InvalidArgument,
+			"provided invalid event",
+			fmt.Errorf(""),
+		)
+	}
+
+	topicHostName, err := s.GetTopicEndpoint(topic)
+	if err != nil {
+		return err
+	}
+	eventToPublish, err := s.NitricEventToEvent(topicHostName, event)
+	if err != nil {
+		return newErr(
+			codes.Internal,
+			"error marshalling event",
+			err,
+		)
+	}
+
+	result, err := s.client.PublishEvents(ctx, topicHostName, eventToPublish)
+	if err != nil {
+		return newErr(
+			codes.Internal,
+			"error publishing event",
+			err,
+		)
+	}
+
+	statusCode := fmt.Sprint(result.StatusCode)
+	if strings.Split(statusCode, "")[0] != "2" {
+		return newErr(
+			codes.Internal,
+			"returned non 200 status code",
+			fmt.Errorf(""),
+		)
+	}
```

Comment:
I'd suggest you don't use a string here to check the status code, keep it as an int. You could also include the status in the error that's returned. E.g.

```go
if result.StatusCode < 200 || result.StatusCode >= 300 {
  return newErr(
    codes.Internal,
    "returned non 200 status code",
    fmt.Errorf(result.Status),
  )
}
```